woocommerce/woocommerce
# Logging: Add a list table to browse log files

### Changes proposed in this Pull Request:

Adds a new file browsing view for Logs and sets the stage for other UI improvements described in #40644. In this view, the list of log files can be sorted by several different parameters and also filtered by the log's "source" (which is essentially the prefix on the log's filename). It has the standard list table feature of being able to select multiple rows at once and perform a bulk action. In this case, the only bulk action is currently to delete files. Clicking on one of the files brings you to the single file view, which at this point is just the old log file view.

Fixes #40915

### Screen shot

![Screen Shot 2023-10-25 at 11 50 06](https://github.com/woocommerce/woocommerce/assets/916023/73ef0b61-b42d-47e3-b6c1-82403bda1719)

### How to test the changes in this Pull Request:

#### Setup

1. After checking this branch out, you'll need to run `composer dump-autoload` since there are new classes that need to be added to the autoloader.
2. In your test site's **wp-config.php** file, add this line: `define( 'WC_LOG_HANDLER', 'Automattic\\WooCommerce\\Internal\\Admin\\Logging\\LogHandlerFileV2' );`. This is what allows you to see the new view.
3. Install [this plugin](https://gist.github.com/coreymckrill/2968911c8b63c556e7787f2cdc65f966) on your test site for generating test log files. Probably easiest to copy the file into the mu-plugins directory.
4. Open two browser tabs: the Logs screen (WP Admin > WooCommerce > Status > Logs) and the WooCommerce Tools screen (WP Admin > WooCommerce > Status > Tools). On the Tools screen you should see the **Debug Log Generator** tool at the top of the list.

#### Tests

1. On the Logs screen, you should see a list table, which may already have some log files listed in it. Even if there are some, go ahead and generate some additional logs on the Tools screen so that we have enough different log files and sources.
2. Try clicking on the column headers to sort the log files by different parameters.
3. Choose a source from the "All Sources" dropdown and click Filter. The list table should now only display log files from that source. Sorting the files via the column headers again should maintain the filter, while still sorting accurately.
4. After using the Debug Log Generator tool, you should have at least three different log files. Click the Screen Options tab at the top of the Logs screen and change the "Number of items per page" field to a small number, such as `2`, so you can test pagination. After applying this change, try navigating between pages in the list table. Filters and sorting should be maintained when you do this.
5. Also on the Screen Options tab, you can try toggling various columns on/off.
6. View the contents of a log file by clicking on its link in the "Source" column. This should take you to a view of that log file, which is also the current production UI for logs when using the file handler (as opposed to the database handler). Improvements to this view will be made in a separate PR, so the important thing for now is just that you can reach this view and see the contents of the file.
7. Back on the list table view, try deleting multiple log files. Check the boxes next to a couple of them and then choose "Delete permanently" from the "Bulk actions" dropdown, and click "Apply".
8. The current handler that generates log files has a mechanism for "rotating" log files, so that any one log file doesn't get too big. It does this by renaming a file that has reached the threshold with a suffix like`.0` at the end of the source segment of the filename. To see how this is represented in the list table, try generating two or three batches of test logs at once. You should then see that at least one of the logs, probably `debug-log-generator`, has more than one entry in the list table. The one with the most recently modified date should just have the source name, but any others should be followed by an N dash and then a number (`- 6`). There's something buggy going on with the way log rotation works, so the rotation suffix may not be in numerical order (i.e. you may have a `6` but not any of the other numbers preceding that). I haven't dug into that bug yet, but it's not related to this PR as far as I can tell, so we'll figure it out later.

### Changelog entry




<details>

#### Significance




#### Type




#### Message

Add a view to the Logs screen for browsing log files.

#### Comment 

</details>
